### PR TITLE
Fix #8690: the signature for product should come from expected type

### LINKED ISSUE #8690: Spurious exhaustiveness warning for union of tuple types
## Minimized code

```Scala
class A
class B

def test(x: (A, B) | (B, A)) = x match {
  case (u: A, v) => (u, v)
  case (u: B, v) => (v, u)
}
```

https://scastie.scala-lang.org/ifiEqLPETBq2eruEbFXfFQ

## Output

```scala
match may not be exhaustive.
It would fail on pattern case: Tuple2(_, _)
```

Moreover, the type inferred for `v` is always `Object`, whereas it could be more precise here.

## Expectation

No warning, and more precise types.

### PR DESCRIPTION
**Pull request by liufengyun**
Fix #8690: the signature for product should come from expected type

### CODE REVIEW

**Comment by liufengyun** [comment]
test performance with #fast please

**Comment by liufengyun** [comment]
test performance please

**Comment by liufengyun** [comment]
test performance please

File: compiler/src/dotty/tools/dotc/transform/patmat/Space.scala [hunk 1]
```diff
@@ -480,15 +502,31 @@ class SpaceEngine(implicit ctx: Context) extends SpaceLogic {
   def isEqualType(tp1: Type, tp2: Type): Boolean = tp1 =:= tp2
 
   /** Parameter types of the case class type `tp`. Adapted from `unapplyPlan` in patternMatcher  */
-  def signature(unapp: Type, unappSym: Symbol, argLen: Int): List[Type] = {
+  def signature(unapp: TermRef, scrutineeTp: Type, argLen: Int): List[Type] = {
+    val unappSym = unapp.symbol
     def caseClass = unappSym.owner.linkedClass
 
+    // println("scrutineeTp = " + scrutineeTp.show)
```
Comment (nicolasstucki) [inline compiler/src/dotty/tools/dotc/transform/patmat/Space.scala:509]: Forgot to remove this

**Pull request merged**